File: src/csharp/howto/HowTo_EngineUse.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

using System;
using System.Collections.Generic;
using System.IO;
using Pulse;
using Pulse.CDM;

namespace HowTo_UseEngine
{
  public class MyEventHandler : IEventHandler
  {
    public void HandleEvent(SEEventChange change)
    {
      if (change.Event == eEvent.HypovolemicShock && change.Active)
        Console.WriteLine(change.ToString());
    }
  }

  public class MyLogListener : ILogListener
  {
    public void Debug(string msg) { Console.WriteLine("[DEBUG] " + msg); }
    public void Info(string msg) { Console.WriteLine("[INFO] " + msg); }
    public void Warning(string msg) { Console.WriteLine("[WARN] " + msg); }
    public void Error(string msg) { Console.WriteLine("[ERROR] " + msg); }
    public void Fatal(string msg) { Console.WriteLine("[FATAL] " + msg); }
  }

  class Example
  {
    protected enum InitializationType { PatientObject, PatientFile, StateFileName, StateString };

    public static void Run()
    {
      Console.WriteLine("Using Pulse version " + PulseEngine.Version() + "-" + PulseEngine.Hash());
      // Here is an array to store Event Changes
      List<SEEventChange> event_changes = new List<SEEventChange>();
      // Here is an array to store Active Events
      List<SEActiveEvent> active_events = new List<SEActiveEvent>();
      // Create a list of Data Requests to specify all the data we want from Pulse
      List<SEDataRequest> data_requests = new List<SEDataRequest>
      {
        // Vitals Monitor Data
        SEDataRequest.CreatePhysiologyDataRequest("HeartRate", FrequencyUnit.Per_min),
        SEDataRequest.CreatePhysiologyDataRequest("ArterialPressure", PressureUnit.mmHg),
        SEDataRequest.CreatePhysiologyDataRequest("MeanArterialPressure", PressureUnit.mmHg),
        SEDataRequest.CreatePhysiologyDataRequest("SystolicArterialPressure", PressureUnit.mmHg),
        SEDataRequest.CreatePhysiologyDataRequest("DiastolicArterialPressure", PressureUnit.mmHg),
        SEDataRequest.CreatePhysiologyDataRequest("OxygenSaturation"),
        SEDataRequest.CreatePhysiologyDataRequest("EndTidalCarbonDioxidePressure", PressureUnit.mmHg),
        SEDataRequest.CreatePhysiologyDataRequest("RespirationRate", FrequencyUnit.Per_min),
        SEDataRequest.CreatePhysiologyDataRequest("SkinTemperature", TemperatureUnit.C),
        SEDataRequest.CreateGasCompartmentDataRequest("Carina", "CarbonDioxide", "PartialPressure", PressureUnit.mmHg),
        SEDataRequest.CreatePhysiologyDataRequest("BloodVolume", VolumeUnit.mL),
        SEDataRequest.CreateECGDataRequest("Lead3ElectricPotential", ElectricPotentialUnit.mV),
      };
      SEDataRequestManager data_mgr = new SEDataRequestManager(data_requests);
      // In addition to getting this data back via this API
      // You can have Pulse write the data you have requested to a CSV file
      data_mgr.SetResultsFilename("./test_results/HowTo_EngineUse.cs.csv");
      // Create a reference to a double[] that will contain the data returned from Pulse
      double[] data_values;
      // data_values[0] is ALWAYS the simulation time in seconds
      // The rest of the data values are in order of the data_requests list provided

      // Instantiate a Pulse engine
      String here = System.IO.Directory.GetCurrentDirectory();
      PulseEngine pulse = new PulseEngine();
      // Ask Pulse for any information you might want
      double ts = pulse.GetTimeStep_s();

      // You can ask Pulse to write out a log file if you want
      // By default, no log file is written
      pulse.SetLogFilename("./test_results/HowTo_EngineUse.cs.log");
      // You can also have pulse log to the console (std::cout)
      // By default, console logging is turned off
      pulse.LogToConsole(false);// We have a logger writing to the console in this example
      // It is recommended to listen and  manage the log messages yourself
      // With a listener, you can add messages to a log, display, or write them to console
      pulse.SetLogListener(new MyLogListener());
      // Listen for events, this is optional
      // You must provide an event listener to get events
      pulse.SetEventHandler(new MyEventHandler());

      InitializationType initType = InitializationType.StateFileName;
      switch (initType)
      {
        case InitializationType.StateFileName:
          {
            // Load a state file
            if (!pulse.SerializeFromFile("./states/Soldier@0s.pbb", data_mgr))
            {
              Console.WriteLine("Error Initializing Pulse!");
              return;
            }
            break;
          }
        case InitializationType.StateString:
          {
            string file_content = File.ReadAllText("./states/Soldier@0s.json");
            if (!pulse.SerializeFromString(file_content, data_mgr, eSerializationFormat.JSON))
            {
              Console.WriteLine("Error Initializing Pulse!");
              return;
            }
            break;
          }
        // Both of these cases will use the SEPatientConfiguration object
        case InitializationType.PatientFile:
          {
            // If you have a patient file, simply load up the file
            SEPatientConfiguration cfg = new SEPatientConfiguration();
            // If we want to specify a file on disk that the C++ has access to, we can specify that file
            // (Maybe the Pulse engine is on hosted on another machine)
            //cfg.SetPatientFile("./patients/Soldier.json");
            // Or, ou can load up a local file on disk
            cfg.GetPatient().SerializeFromFile("./patients/Soldier.json");
            // Optionally, you can add conditions to the patient
            cfg.GetConditions().GetAcuteRespiratoryDistressSyndrome().GetSeverity().SetValue(0.2);
            cfg.GetConditions().GetAcuteRespiratoryDistressSyndrome().GetLeftLungAffected().SetValue(1.0);
            cfg.GetConditions().GetAcuteRespiratoryDistressSyndrome().GetRightLungAffected().SetValue(1.0);
            // Initialization requires that you have all data files on disk for Pulse to find when creating a patient
            //   - Patient file (if not creating an inline patient definition)
            //   - Substance, Stabilization (config folder), nutrition, environments, ecg files
            // The data_dir field can be changed in the InitializeEngine method
            // If you have those folders in another location
            if (!pulse.InitializeEngine(cfg, data_mgr))
            {
              Console.WriteLine("Error Initializing Pulse!");
              return;
            }
            // This takes a little while for the engine to find the patient homeostatic/stable state
            break;
          }
        case InitializationType.PatientObject:
          {
            SEPatientConfiguration cfg = new SEPatientConfiguration();
            // Grab the patient and fill in some data
            SEPatient patient = cfg.GetPatient();
            patient.SetName("Owen");
            patient.SetSex(ePatient_Sex.Male);
            patient.GetAge().SetValue(30, TimeUnit.yr);
            patient.GetWeight().SetValue(200, MassUnit.lb);
            patient.GetHeight().SetValue(74, LengthUnit.inch);
            patient.GetBodyFatFraction().SetValue(0.16);
            patient.GetSystolicArterialPressureBaseline().SetValue(120, PressureUnit.mmHg);
            patient.GetDiastolicArterialPressureBaseline().SetValue(72, PressureUnit.mmHg);
            patient.GetHeartRateBaseline().SetValue(67, FrequencyUnit.Per_min);
            patient.GetRespirationRateBaseline().SetValue(12, FrequencyUnit.Per_min);
            // You really only need to set the name
            // All other patient values will be computed based on
            // https://pulse.kitware.com/_patient_methodology.html
            // Optionally, you can add conditions to the patient
            cfg.GetConditions().GetChronicObstructivePulmonaryDisease().GetBronchitisSeverity().SetValue(0.45);
            cfg.GetConditions().GetChronicObstructivePulmonaryDisease().GetEmphysemaSeverity().SetValue(0.20);
            // Initialization requires that you have all data files on disk for Pulse to find when creating a patient
            //   - Patient file (if not creating an inline patient definition)
            //   - Substance, Stabilization (config folder), nutrition, environments, ecg files
            // If you have those folders in another location,
            // The data root dir can be changed in the SEPatientConfiguration object
            cfg.SetDataRootDir("./");// The default is "./"
            if (!pulse.InitializeEngine(cfg, data_mgr))
            {
              Console.WriteLine("Error Initializing Pulse!");
              return;
            }
            // This takes a little while for the engine to find the patient homeostatic/stable state
            break;
          }
      }
      // DOES ANYONE WANT TO BE ABLE TO CHANGE DATA REQUESTS IN THE MIDDLE OF A RUN?
      // NOTE ANY CSV FILE BEING WRITTEN OUT WOULD NOT SUPPORT CHANGING DATA IN THE MIDDLE OF A RUN
      // BUT IF YOU ARE NOT WRITING A CSV OUT, I COULD SEE THIS BEING USEFUL...

      // You can get the initial patient at any time
      // But it will not change, so once is good
      // All values will be set to what the engine stabilized to
      SEPatient initialPatient = new SEPatient();
      pulse.GetInitialPatient(initialPatient);
      Console.WriteLine("Sex " + initialPatient.GetSex().ToString());
      Console.WriteLine("Age " + initialPatient.GetAge().ToString());
      Console.WriteLine("Height " + initialPatient.GetHeight().ToString());
      Console.WriteLine("Weight " + initialPatient.GetWeight().ToString());

      // You can pull the condition manager from the engine to see what, if any, conditions are applied to this patient
      // Note this will not change, its based on how the patient was initially created
      List<SECondition> conditions = new List<SECondition>();
      pulse.GetConditions(conditions);
      foreach (SECondition any in conditions)
      {
        Console.WriteLine(any.ToString());
      }

      // Now we can start telling the engine what to do
      // All the same concepts apply from the C++ HowTo files, so look there if you want to see more examples

      // Advance time and print out values
      for (int i = 1; i <= 10; i++)
      {
        if (!pulse.AdvanceTime_s(0.65))
          Console.WriteLine("Error Advancing Time!");
        else
        {
          // Pull data from pulse
          data_values = pulse.PullData();
          // And write it out to the console
          data_mgr.WriteData(data_values);

          // The engine can provide you with all the current
          // active events and how long they have been active for
          // Active events are reevaluated each call to PullActiveEvents
          pulse.PullActiveEvents(active_events);
          foreach (var active_event in active_events)
          {
            Console.WriteLine(active_event.ToString());
          }
        }
      }

      // At any time you can get a patient assessment from the engine
      SECompleteBloodCount cbc = new SECompleteBloodCount();
      pulse.GetPatientAssessment(cbc);

      // Let's do something to the patient, you can either send actions over one at a time, or pass in a List<SEAction>
      List<SEAction> actions = new List<SEAction>();

      SEHemorrhage h = new SEHemorrhage();
      h.SetExternal(SEHemorrhage.ExternalCompartment.RightLeg);
      // Optionally, You can set the flow rate of the hemorrhage,
      // This needs to be provided the proper flow rate associated with the anatomy
      // This is implemented as a flow source, this rate will be constant, and will not be affected by dropping blood pressures
      // It is intended to interact with sensors or with something continuously monitoring physiology and updating the flow
      //h.GetFlowRate().SetValue(50, VolumePerTimeUnit.mL_Per_min);// Change this to 750 if you want to see how engine failures are handled!!
      actions.Add(h);
      if (!pulse.ProcessActions(actions))
      {
        Console.WriteLine("Engine was unable to process requested actions");
        return;
      }
      // Note CDM is not updated after this call, you have to advance some time
      for (int i = 1; i <= 2; i++)
      {
        if (!pulse.AdvanceTime_s(60)) // Simulate one minute
        {
          Console.WriteLine("Engine was unable to stay within modeling parameters with requested actions");
          return;
        }
        // Get the values of the data you requested at this time
        data_values = pulse.PullData();
        // And write it out to the console
        data_mgr.WriteData(data_values);
      }

      // At any point you can also pull the current active actions from the engine
      Console.WriteLine("Here are the current active actions in the engine");
      pulse.PullActiveActions(actions);
      foreach (SEAction any in actions)
      {
        Console.WriteLine(any.ToString());
      }

      // Stop the hemorrhage
      h.GetSeverity().SetValue(0);
      //h.GetFlowRate().SetValue(0, VolumePerTimeUnit.mL_Per_min);
      if (!pulse.ProcessAction(h))
      {
        Console.WriteLine("Engine was unable to process requested actions");
        return;
      }

      for (int i = 1; i <= 1; i++)
      {
        if (!pulse.AdvanceTime_s(60)) // Simulate one minute
        {
          Console.WriteLine("Engine was unable to stay within modeling parameters with requested actions");
          return;
        }
        // Pull data from pulse
        data_values = pulse.PullData();
        // And write it out to the console
        data_mgr.WriteData(data_values);
      }

      // Administer Drugs
      SESubstanceInfusion paralytic = new SESubstanceInfusion();
      paralytic.SetSubstance("Succinylcholine");
      paralytic.GetConcentration().SetValue(5000, MassPerVolumeUnit.ug_Per_mL);
      paralytic.GetRate().SetValue(100, VolumePerTimeUnit.mL_Per_min);
      if (!pulse.ProcessAction(paralytic))
      {
        Console.WriteLine("Engine was unable to process requested actions");
        return;
      }
      //Let's use the anesthesia machine to simulate an ambu bag for ventilation
      // We could use the mechanical ventilation action, but you have to provide a changing pressure per squeeze
      // And I just want things automatic for this demo
      // But if you want to interact with a haptic device, use that action with data from your device
      // Or program up a sinusoidal or square wave for applying pressure
      SEAnesthesiaMachineConfiguration am = new SEAnesthesiaMachineConfiguration();
      am.GetConfiguration().SetConnection(eSwitch.On);
      am.GetConfiguration().GetInletFlow().SetValue(5, VolumePerTimeUnit.L_Per_min);
      am.GetConfiguration().GetInspiratoryExpiratoryRatio().SetValue(0.5);
      am.GetConfiguration().GetOxygenFraction().SetValue(0.23);
      am.GetConfiguration().SetOxygenSource(eAnesthesiaMachine_OxygenSource.Wall);
      am.GetConfiguration().GetPeakInspiratoryPressure().SetValue(10.5, PressureUnit.cmH2O);
      am.GetConfiguration().GetPositiveEndExpiredPressure().SetValue(1, PressureUnit.cmH2O);
      am.GetConfiguration().SetPrimaryGas(eAnesthesiaMachine_PrimaryGas.Nitrogen);
      am.GetConfiguration().GetRespiratoryRate().SetValue(16, FrequencyUnit.Per_min);
      if (!pulse.ProcessAction(am))
      {
        Console.WriteLine("Engine was unable to process requested actions");
        return;
      }

      // Infuse some fluids
      SESubstanceCompoundInfusion ivFluids = new SESubstanceCompoundInfusion();
      ivFluids.SetSubstanceCompound("Saline");
      ivFluids.GetBagVolume().SetValue(500, VolumeUnit.mL);
      ivFluids.GetRate().SetValue(100, VolumePerTimeUnit.mL_Per_min);
      if (!pulse.ProcessAction(ivFluids))
      {
        Console.WriteLine("Engine was unable to process requested actions");
        return;
      }

      for (int i = 1; i <= 5; i++)
      {
        if (!pulse.AdvanceTime_s(60)) // Simulate one minute
        {
          Console.WriteLine("Engine was unable to stay within modeling parameters with requested actions");
          return;
        }
        // Pull data from pulse
        data_values = pulse.PullData();
        // And write it out to the console
        data_mgr.WriteData(data_values);
      }
    }
  }
}

File: src/java/com/kitware/pulse/howto/HowTo_EngineUse.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
package com.kitware.pulse.howto;

import java.util.ArrayList;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.kitware.pulse.cdm.bind.Engine.DataRequestData.eCategory;
import com.kitware.pulse.cdm.bind.Events.eEvent;
import com.kitware.pulse.cdm.bind.Patient.PatientData.eSex;
import com.kitware.pulse.cdm.bind.PatientActions.HemorrhageData;
import com.kitware.pulse.cdm.conditions.SECondition;
import com.kitware.pulse.cdm.datarequests.SEDataRequest;
import com.kitware.pulse.cdm.datarequests.SEDataRequestManager;
import com.kitware.pulse.cdm.engine.SEActiveEvent;
import com.kitware.pulse.cdm.engine.SEEventHandler;
import com.kitware.pulse.cdm.engine.SEPatientConfiguration;
import com.kitware.pulse.cdm.patient.SEPatient;
import com.kitware.pulse.cdm.patient.actions.SEHemorrhage;
import com.kitware.pulse.cdm.patient.actions.SESubstanceCompoundInfusion;
import com.kitware.pulse.cdm.patient.assessments.SECompleteBloodCount;
import com.kitware.pulse.cdm.patient.conditions.SEChronicAnemia;
import com.kitware.pulse.cdm.properties.CommonUnits.FrequencyUnit;
import com.kitware.pulse.cdm.properties.CommonUnits.LengthUnit;
import com.kitware.pulse.cdm.properties.CommonUnits.MassUnit;
import com.kitware.pulse.cdm.properties.CommonUnits.PressureUnit;
import com.kitware.pulse.cdm.properties.CommonUnits.TimeUnit;
import com.kitware.pulse.cdm.properties.CommonUnits.VolumePerTimeUnit;
import com.kitware.pulse.cdm.properties.CommonUnits.VolumeUnit;
import com.kitware.pulse.cdm.properties.SEScalarTime;
import com.kitware.pulse.engine.PulseCompartments;
import com.kitware.pulse.engine.PulseEngine;
import com.kitware.pulse.utilities.Log;
import com.kitware.pulse.utilities.LogListener;
import com.kitware.pulse.utilities.jniBridge;

/**
 * How to use the Java interface to the Pulse Physiology Engine
 */
public class HowTo_EngineUse
{
 // Create a listener that will catch any messages logged in C++
 // This class will take the messages and add them to the log created in Java
 protected static class MyListener extends LogListener
 {
   protected MyListener()
   {
     super();
     // Since we are just pushing this data into the log
     // We set listen to false, when you create a LogListener
     // is will automatically register itself with the static Log class,
     // and whenever a Log even class is called, this will be called
     // and since this just calls the Log to log you will get into a recursive infinite loop.
     // This is just because I use the LogListener interface to listen to any log messages coming
     // from C++. Technically it is a LogListener as it is 'listening' to log events in C++.
     // At any rate the Java side will receive log events from C++, and you create your own 
     // LogListener so you can do as you see fit with those messages. BUT if you do want to 
     // push those messages to the Java Log, you will need this class to not listen so you
     // don't get into an infinite recursive loop
     listen(false);
   }    
   @Override public void handleDebug(String msg) { Log.debug(msg); }
   @Override public void handleInfo(String msg)  { Log.info(msg); }
   @Override public void handleWarn(String msg)  { Log.warn(msg); }
   @Override public void handleError(String msg) { Log.error(msg); }
   @Override public void handleFatal(String msg) { Log.fatal(msg); }
 }
 
 // Another listener example
 // Create a listener that will catch any messages logged in C++
 // This class will take the messages and forward them to slf4j,
 // so they can be logged with whatever logger a user may want to use
 protected static class ForwardToSlf4jListener extends LogListener
 {
   private static final Logger LOG = LoggerFactory.getLogger("CppPulseEngine"); // select a name that makes it clear where these logs are coming from
   
   @Override public void handleDebug(String msg) { LOG.debug(msg); }
   @Override public void handleDebug(String msg, Throwable t) { LOG.debug(msg, t); }
   @Override public void handleInfo(String msg)  { LOG.info(msg); }
   @Override public void handleInfo(String msg, Throwable t)  { LOG.info(msg, t); }
   @Override public void handleWarn(String msg)  { LOG.warn(msg); }
   @Override public void handleWarn(String msg, Throwable t)  { LOG.warn(msg, t); }
   @Override public void handleError(String msg) { LOG.error(msg); }
   @Override public void handleError(String msg, Throwable t) { LOG.error(msg, t); }
   @Override public void handleFatal(String msg) { LOG.error("FATAL: {}", msg); }
   @Override public void handleFatal(String msg, Throwable t) { LOG.error("FATAL: {}", msg, t); }
 }
 
 protected static class MyEventHandler implements SEEventHandler
 {
  @Override
  public void handleEvent(eEvent e, boolean active, SEScalarTime time)
  {
    // Here is how you test against an event of interest and do something
   if(e == eEvent.CardiacArrest && active)
      Log.info("Patient is in cardiac arrest "+time);
  }
 }
 
 public enum InitializationType { PatientObject, PatientFile, StateFile };
 
 public static void main(String[] args)
 {
   example();
 }
 
 public static void example()
 {
   // As Pulse is a C++ library, a JNI bridge is necessary to pass data between C++ and Java
   // Pulse is contained in a single dynamic/shared library, that will need to be loaded into the JVM
   // By default an instantiation of the PulseEngine will look for and load the Pulse library in the java.library.path
   
   // If the library is not found in java.library.path,
   // it will look in a directory you provide or user.dir if not directory is provided
   //jniBridge.initialize("/path/to/pulse/library");
   // This should also be called manually if you are using the Pulse API (such as Scalars) without the PulseEngine
   // Currently, the jniBridge.deinitialize() method is not necessary to call
   
   // If you would like to run against debug library (PulseJNId), you can do initialize like this
   //jniBridge.initializeDebug();
   
   // Create a Pulse Engine
   PulseEngine pe = new PulseEngine();
   
   // By default, no log file is written
   pe.setLogFilename("./test_results/HowTo_EngineUse.java.log");
   
   // The Pulse API provides a Log class
   // It can be used to write to easily forward messages to the console and any LogListener
   // So messages from this example, and the Pulse Engine will both goto the same slf4j logger
   // If you use the Log (like this example), it by default outputs logged messages to the console
   // We are using a slf4j listener that is also going to write to the console.
   // We are going to turn off console writing by the Log class, and it will only
   // direct log messages to our Listener, which will write them to the console
   Log.output2Console = false;
   
   // Create a listener that will get any log messages that come from the engine.
   // (Info, Warnings, Errors, Fatal Errors)
   
   // This forwards log messages to an SLF4J logger
   // Note that pulse comes with the logback logger so slf4j will print something for this example
   // If you would rather use a different logger, you can delete the logback jars and substitute something else
   pe.setLogListener(new ForwardToSlf4jListener());
   
   // If you want to know whenever the engine enters and exits a particular state
   pe.setEventHandler(new MyEventHandler());
   
   // Here is how to specify the data to get back from the engine
   SEDataRequestManager dataRequests = new SEDataRequestManager();
   SEDataRequest hr = new SEDataRequest();
   hr.setCategory(eCategory.Physiology);
   hr.setPropertyName("HeartRate");
   hr.setUnit(FrequencyUnit.Per_min.toString());
   dataRequests.getRequestedData().add(hr);
   SEDataRequest rr = new SEDataRequest();
   rr.setCategory(eCategory.Physiology);
   rr.setPropertyName("RespirationRate");
   rr.setUnit(FrequencyUnit.Per_min.toString());
   dataRequests.getRequestedData().add(rr);
   SEDataRequest tlv = new SEDataRequest(); 
   tlv.setCategory(eCategory.Physiology);   
   tlv.setPropertyName("TotalLungVolume");
   tlv.setUnit(VolumeUnit.mL.toString());
   dataRequests.getRequestedData().add(tlv);
   SEDataRequest bv = new SEDataRequest();  
   bv.setCategory(eCategory.Physiology);  
   bv.setPropertyName("BloodVolume");
   bv.setUnit(VolumeUnit.mL.toString());
   dataRequests.getRequestedData().add(bv);
   // In addition to getting this data back via this API
   // You can have Pulse write the data you have requested to a CSV file
   dataRequests.setResultsFilename("./test_results/HowTo_EngineUse.java.csv");
   // Create a reference to a double[] that will contain the data returned from Pulse
   List<Double> dataValues;
   // data_values[0] is ALWAYS the simulation time in seconds
   // The rest of the data values are in order of the data_requests list provided
   
   InitializationType initType = InitializationType.StateFile;
   // INITIALIZE THE ENGINE WITH A PATIENT
   switch(initType)
   {
   case PatientObject:
     {
      // Set up your patient
      // This is the minimum and recommended data to set for a patient
      // But you can get more complicated, consult our Patient Variability Documentation for more details
      SEPatient patient = new SEPatient();
      patient.setName("Standard");
      patient.setSex(eSex.Male);
      patient.getAge().setValue(44,TimeUnit.yr);
      patient.getWeight().setValue(170,MassUnit.lb);
      patient.getHeight().setValue(71,LengthUnit.in);
      patient.getBodyFatFraction().setValue(0.21);
      patient.getHeartRateBaseline().setValue(72,FrequencyUnit.Per_min);
      patient.getDiastolicArterialPressureBaseline().setValue(72.,PressureUnit.mmHg);
      patient.getSystolicArterialPressureBaseline().setValue(114.,PressureUnit.mmHg);
      patient.getRespirationRateBaseline().setValue(16,FrequencyUnit.Per_min);
      
      SEPatientConfiguration patient_configuration = new SEPatientConfiguration();
      patient_configuration.setPatient(patient);
      // Optionally add conditions to the patient_configuration
      
      
      // If I wanted to make set a condition on the patient
      // Note that while you can have multiple conditions on a patient
      // It is more than likely not tested and the engine may or may not converge
      List<SECondition> conditions = new ArrayList<>();
      SEChronicAnemia anemia = new SEChronicAnemia();// One of these days I will prefix these with SE...
      anemia.getReductionFactor().setValue(0.3);
      conditions.add(anemia);
      
      // Initialize an engine 
      // Optionally, initializeEngine can take in a directory where Pulse looks for its required data files on disk
      pe.initializeEngine(patient_configuration, dataRequests);
      //pe.initializeEngine(patient_configuration, dataRequests, "./");
       // This method will block while the engine stabilizes to meet the defined patient parameters
       break;
     }
   case PatientFile:
     {
    	 SEPatientConfiguration patient_configuration = new SEPatientConfiguration();
       patient_configuration.setPatientFile("./patient/StandardMale.json");
       // Optionally add conditions to the patient_configuration
       
       // Allocate an engine
       pe.initializeEngine(patient_configuration, dataRequests);
       // This method will block while the engine stabilizes to meet the defined patient parameters
       break;
     }
   case StateFile:
     {
       pe.serializeFromFile("./states/StandardMale@0s.json", dataRequests);
       // This method method sets the engine to the provided state instantaneously and you are ready to process actions/advance time
       break;
     }
   }
   
   // You can get the initial patient at any time
   // But it will not change, so once is good
   // All values will be set to what the engine stabilized to
   SEPatient initialPatient = new SEPatient();
   pe.getInitialPatient(initialPatient);
   Log.info("Sex " + initialPatient.getSex());
   Log.info("Age(yr) " + initialPatient.getAge().getValue(TimeUnit.yr));
   Log.info("Height(cm) " + initialPatient.getHeight().getValue(LengthUnit.cm));
   Log.info("Weight(lb) " + initialPatient.getWeight().getValue(MassUnit.lb));
   
   // Now we can start telling the engine what to do
   // All the same concepts apply from the C++ HowTo files, so look there if you want to see more examples
   
   if(!pe.advanceTime()) // Advance one engine time step
   {
     Log.info("Something bad happened");
     return;
   }
   
   dataValues = pe.pullData();
   Log.info("Simulation Time(s) " + dataValues.get(0));
   Log.info("Heart Rate(bpm) " + dataValues.get(1));
   Log.info("Respiration Rate(bpm) " + dataValues.get(2));
   Log.info("Total Lung Volume(mL) " + dataValues.get(3));
   Log.info("Blood Volume(mL) " + dataValues.get(4));
   
   SEScalarTime time = new SEScalarTime(0, TimeUnit.s);
   time.setValue(1, TimeUnit.s);
   Log.info("Advancing "+time+"...");
   if(!pe.advanceTime(time)) // Simulate one second
   {
     Log.info("Something bad happened");
     return;
   }
   
   dataValues = pe.pullData();
   Log.info("Simulation Time(s) " + dataValues.get(0));
   Log.info("Heart Rate(bpm) " + dataValues.get(1));
   Log.info("Respiration Rate(bpm) " + dataValues.get(2));
   Log.info("Total Lung Volume(mL) " + dataValues.get(3));
   Log.info("Blood Volume(mL) " + dataValues.get(4));
   
   // Let's do something to the patient, you can either send actions over one at a time, or pass in a List<SEAction>
   SEHemorrhage h = new SEHemorrhage();
   h.setType(HemorrhageData.eType.External);
   h.setCompartment(PulseCompartments.Vascular.RightLeg);
   h.getRate().setValue(75,VolumePerTimeUnit.mL_Per_min);// Change this to 750 if you want to see how engine failures are handled!!
   if(!pe.processAction(h))
   {
     Log.error("Engine was unable to process requested actions");
     return;
   }

   for(int i=1; i<=2; i++)
   {
  	 time.setValue(1,TimeUnit.min);
     Log.info("Advancing "+time+"...");
	   if(!pe.advanceTime(time)) // Simulate one minute
	   {
	     Log.error("Engine was unable to stay within modeling parameters with requested actions");
	     return;
	   }
	   dataValues = pe.pullData();
	   Log.info("Simulation Time(s) " + dataValues.get(0));
	   Log.info("Heart Rate(bpm) " + dataValues.get(1));
	   Log.info("Respiration Rate(bpm) " + dataValues.get(2));
	   Log.info("Total Lung Volume(mL) " + dataValues.get(3));
	   Log.info("Blood Volume(mL) " + dataValues.get(4));
   }
   
   // Stop the hemorrhage
   h.getRate().setValue(0,VolumePerTimeUnit.mL_Per_min);
   if(!pe.processAction(h))
   {
     Log.error("Engine was unable to process requested actions");
     return;
   }
   
   for(int i=1; i<=1; i++)
   {
     time.setValue(1,TimeUnit.min);
     Log.info("Advancing "+time+"...");
     if(!pe.advanceTime(time)) // Simulate one minute
     {
       Log.error("Engine was unable to stay within modeling parameters with requested actions");
       return;
     }
     dataValues = pe.pullData();
     Log.info("Simulation Time(s) " + dataValues.get(0));
     Log.info("Heart Rate(bpm) " + dataValues.get(1));
     Log.info("Respiration Rate(bpm) " + dataValues.get(2));
     Log.info("Total Lung Volume(mL) " + dataValues.get(3));
     Log.info("Blood Volume(mL) " + dataValues.get(4));
   }
   
   // Infuse some fluids
   SESubstanceCompoundInfusion ivFluids= new SESubstanceCompoundInfusion();
   ivFluids.setSubstanceCompound("Saline");
   ivFluids.getBagVolume().setValue(500,VolumeUnit.mL);
   ivFluids.getRate().setValue(100,VolumePerTimeUnit.mL_Per_min);
   if(!pe.processAction(ivFluids))
   {
     Log.error("Engine was unable to process requested actions");
     return;
   }
   
   for(int i=1; i<=5; i++)
   {
     time.setValue(1,TimeUnit.min);
     Log.info("Advancing "+time+"...");
     if(!pe.advanceTime(time)) // Simulate one minute
     {
       Log.error("Engine was unable to stay within modeling parameters with requested actions");
       return;
     }
     dataValues = pe.pullData();
     Log.info("Simulation Time(s) " + dataValues.get(0));
     Log.info("Heart Rate(bpm) " + dataValues.get(1));
     Log.info("Respiration Rate(bpm) " + dataValues.get(2));
     Log.info("Total Lung Volume(mL) " + dataValues.get(3));
     Log.info("Blood Volume(mL) " + dataValues.get(4));
   }
   
   // You can explicitly check if the patient is in a specific state/event
   List<SEActiveEvent> activeEvents = new ArrayList<SEActiveEvent>();
   pe.getActiveEvents(activeEvents);
   for(SEActiveEvent ae : activeEvents)
     Log.info(ae.type+" has been active for "+ae.duration);
   
   // Let's get an assessment from the engine
   // Assessments can involve extra calculation to generate the data necessary for the specified assessment
   SECompleteBloodCount cbc = new SECompleteBloodCount();
   if(!pe.getPatientAssessment(cbc))
     Log.error("Failed to get patient assessment");
   Log.info("Red Blood Count "+cbc.getRedBloodCellCount());
   Log.info("White Blood Count "+cbc.getWhiteBloodCellCount());
   
   // Be nice to your memory and deallocate the native memory associated with this engine if you are done with it
   pe.cleanUp();
   // Note you can now run a static (scenario) or another dynamic engine with the pe object, it will allocate and manage a new C++ engine 
 }
}
